File: 2018/python/day13.py
class cart(object):
  def __init__(self, id, dir, x, y, currTrack):
    self.id = id
    self.dir = dir
    self.nextTurn = "left"
    self.currTrack = currTrack
    self.x = x
    self.y = y
    self.crashed = False

  def changeDir(self):
    dr = dirRules[self.dir]
    if self.currTrack == "+":
      self.dir = dr.intersection[self.nextTurn]
      self.nextTurn = turningOrder[self.nextTurn]
    else:
      self.dir = dr.dirChange[self.currTrack]

class directionRule(object):
  def __init__(self, xOffset, yOffset, dirChange, intersection):
    self.xOffset = xOffset
    self.yOffset = yOffset
    self.dirChange = dirChange
    self.intersection = intersection

cartParseRules = {"^":"|", ">":"-", "v":"|", "<":"-"}

dirRules = {"^":directionRule(0, -1, {"/":">", "|":"^", "\\":"<", "-":"X"}, {"left":"<","straight":"^","right":">"}),
            "v":directionRule(0,  1, {"/":"<", "|":"v", "\\":">", "-":"X"}, {"left":">","straight":"v","right":"<"}),
            "<":directionRule(-1, 0, {"/":"v", "|":"X", "\\":"^", "-":"<"}, {"left":"v","straight":"<","right":"^"}),
            ">":directionRule( 1, 0, {"/":"^", "|":"X", "\\":"v", "-":">"}, {"left":"^","straight":">","right":"v"})}

turningOrder = {"left":"straight", "straight":"right", "right":"left"}

def parseInput(tracks, carts):
  f = open("day13/input", "r")
  lines = f.read().splitlines()
  for y, line in enumerate(lines):
    row = []
    for x, char in enumerate(line):
      if char in cartParseRules:
        c = cart(len(carts), char, x, y, cartParseRules[char])
        carts.append(c)
        row.append(str(c.id))
      else:
        row.append(char)
    tracks.append(row)
  return tracks

def run(tracks, carts):
  while len(carts) > 1:
    carts.sort(key=lambda c: (c.y, c.x))
    for c in carts:
      if c.crashed:
        continue
      newX = c.x+dirRules[c.dir].xOffset
      newY = c.y+dirRules[c.dir].yOffset
      if [newX, newY] in ([c.x, c.y] for c in carts if not c.crashed):
        otherCart = [c for c in carts if not c.crashed and c.x == newX and c.y == newY][0]
        print("BOOM!! @ %d,%d" % (newX, newY))
        tracks[otherCart.y][otherCart.x] = otherCart.currTrack
        otherCart.crashed = True
        tracks[c.y][c.x] = c.currTrack
        c.crashed = True
      else: 
        tracks[c.y][c.x] = c.currTrack
        c.currTrack = tracks[newY][newX]
        tracks[newY][newX] = str(c.id)
        c.x = newX
        c.y = newY
        c.changeDir()

    carts = [c for c in carts if not c.crashed]

  if len(carts) == 1:
    print("Last cart @ %d,%d" % (carts[0].x, carts[0].y))

carts = []
tracks = []
parseInput(tracks, carts)
run(tracks, carts)

File: 2018/go/src/day13/day13.go
package main

import (
	"fmt"
	"sort"
	"utils"
)

type cart struct {
	x        int
	y        int
	dir      rune
	nextTurn rune
	crashed  bool
}

var cartStartTrack = map[rune]rune{
	'^': '|', '<': '-', 'v': '|', '>': '-',
}
var nextTurn = map[rune]rune{
	'l': 's', 's': 'r', 'r': 'l',
}
var directionOffset = map[rune][]int{
	'^': {0, -1}, '<': {-1, 0}, 'v': {0, 1}, '>': {1, 0},
}
var directionChange = map[rune]map[rune]rune{
	'^': {'/': '>', '|': '^', '\\': '<', '-': 'X'},
	'v': {'/': '<', '|': 'v', '\\': '>', '-': 'X'},
	'<': {'/': 'v', '|': 'X', '\\': '^', '-': '<'},
	'>': {'/': '^', '|': 'X', '\\': 'v', '-': '>'},
}
var intersectionDirection = map[rune]map[rune]rune{
	'^': {'l': '<', 's': '^', 'r': '>'},
	'v': {'l': '>', 's': 'v', 'r': '<'},
	'<': {'l': 'v', 's': '<', 'r': '^'},
	'>': {'l': '^', 's': '>', 'r': 'v'},
}

func parseInputToTracksAndCarts(lines []string) (tracks [][]rune, carts []cart) {
	for y, l := range lines {
		var row []rune
		for x, r := range l {
			if startTrack, ok := cartStartTrack[r]; ok {
				c := cart{x, y, r, 'l', false}
				carts = append(carts, c)
				row = append(row, startTrack)
			} else {
				row = append(row, r)
			}
		}
		tracks = append(tracks, row)
	}
	return
}

func find(carts []cart, x int, y int) int {
	for i, c := range carts {
		if c.x == x && c.y == y {
			return i
		}
	}
	return -1
}

func run(tracks [][]rune, carts []cart) {
	for len(carts) > 1 {
		sort.Slice(carts, func(i, j int) bool {
			if carts[i].y != carts[j].y {
				return carts[i].y < carts[j].y
			}
			return carts[i].x < carts[j].x
		})

		for ci, c := range carts {
			if c.crashed {
				continue
			}
			newX := c.x + directionOffset[c.dir][0]
			newY := c.y + directionOffset[c.dir][1]
			if i := find(carts, newX, newY); i != -1 {
				fmt.Printf("BOOM! @ %d,%d\n", newX, newY)
				carts[i].crashed = true
				carts[ci].crashed = true
			} else { // move the cart
				carts[ci].x = newX
				carts[ci].y = newY
				if t := tracks[newY][newX]; t == '+' {
					carts[ci].dir = intersectionDirection[c.dir][c.nextTurn]
					carts[ci].nextTurn = nextTurn[c.nextTurn]
				} else {
					carts[ci].dir = directionChange[c.dir][t]
				}
			}
		}

		i := 0
		for _, c := range carts {
			if !c.crashed {
				carts[i] = c
				i++
			}
		}
		carts = carts[:i]

		if len(carts) == 1 {
			fmt.Printf("Last cart: %v\n", carts[0])
		}
	}
}

func main() {
	lines := utils.ReadFileToLines("day13/input")
	tracks, carts := parseInputToTracksAndCarts(lines)
	run(tracks, carts)
}
